File: server/work/activity.go
package work

import (
	"log/slog"

	"github.com/pockode/server/session"
)

// Activity is what a work item is doing, as one value. It is derived, never
// stored: a work record says what the engine may do with the item (Status) and
// what it is waiting for (Wait), and the session it runs in says what is
// happening right now. Nothing on screen may spell a state out of those parts
// itself — see docs/lifecycle-ui.md.
//
// Eight leaves, and there is never more than one: the layers it is derived from
// are each exclusive.
//
// It is not the whole of what a row draws. "This work has questions waiting for
// an answer" is a second, independent dimension — an agent that posted one and
// carried on is *running* and has something for the user — and it is counted
// beside this rather than folded into it (RowState).
type Activity string

const (
	// ActivityOpen is a work that exists and was never started.
	ActivityOpen Activity = "open"
	// ActivityRunning is a turn producing output.
	ActivityRunning Activity = "running"
	// ActivityNeedsPermission is a turn blocked on a permission request.
	ActivityNeedsPermission Activity = "needs_permission"
	// ActivityBackground is a turn parked on work that outlives the tool call
	// that started it. Nothing for the user to do, and nothing is stuck.
	ActivityBackground Activity = "background"
	// ActivityWaitingChildren is a story waiting for its subtasks to close.
	ActivityWaitingChildren Activity = "waiting_children"
	// ActivityIdle is an engine-driven work with nothing happening in it right
	// now — between turns, or waiting for the nudge that follows one.
	ActivityIdle Activity = "idle"
	// ActivityStopped is a work the engine does not touch until a person acts.
	ActivityStopped Activity = "stopped"
	// ActivityClosed is a finished work.
	ActivityClosed Activity = "closed"
)

// NeedsUser reports whether the user is the one holding this work up. It is
// exactly one leaf now — a permission request — because a background wait and a
// wait on children are things happening rather than things to do, and a question
// an agent posted holds nothing up at all.
//
// It is therefore not the whole of the attention dot: a work with unanswered
// questions needs the user too, and no activity can say so (see RowState). The
// dot is the one place the two dimensions are merged, and that merge belongs to
// whatever is drawing it, not here.
func (a Activity) NeedsUser() bool {
	return a == ActivityNeedsPermission
}

// DeriveActivity is the rule, and the whole of it:
//
//	The session says what is happening; the work's wait says what it is waiting
//	for when nothing is happening.
//
// turn is the state of the session the work runs in, or the zero value when it
// has no session — which reads as idle, the same as a session sitting between
// turns, because for a work item the two are the same nothing.
//
// Phase outranks wait rather than the other way round: a wait is a standing
// intention, a phase is a fact about this second. An agent that calls story_wait
// and then keeps writing for another ten seconds *is* running, and the row
// should say so; the moment the turn settles the wait takes over. The
// alternative needs a priority table between two kinds of waiting that can
// legitimately coexist, and every entry in such a table is an arbitrary choice
// someone later "fixes".
//
// The client has the same function for sessions (web/src/lib/activity.ts). The
// two are checked against one shared table of cases — work/testdata/
// activity_cases.json — so that the rule has one definition even though it is
// evaluated in two places, for the reason docs/lifecycle-ui.md §1.3 gives: a
// work list spans worktrees, and a client cannot hold the turn state of a
// session in a worktree it has not opened.
func DeriveActivity(w Work, turn session.TurnState) Activity {
	switch w.Status {
	case StatusOpen:
		return ActivityOpen
	case StatusClosed:
		return ActivityClosed
	case StatusStopped:
		return ActivityStopped
	}

	switch turn.Phase {
	case session.PhaseRunning:
		return ActivityRunning
	case session.PhaseBlocked:
		return blockedActivity(turn)
	}

	// Idle, or no session at all.
	if w.Wait == WaitChild {
		return ActivityWaitingChildren
	}
	return ActivityIdle
}

// blockedActivity picks the leaf for a blocked turn. Permission outranks
// background because background is the one nobody can act on — and those are
// the only two blockers left, now that a question belongs to the session rather
// than to a turn.
func blockedActivity(turn session.TurnState) Activity {
	for _, b := range turn.Blockers {
		if b.Kind == session.BlockerPermission {
			return ActivityNeedsPermission
		}
	}
	return ActivityBackground
}

// RowState is everything a work row draws that the work record does not know:
// what the item is doing, and how many of its questions are waiting for the
// user.
//
// The two travel together because they are read from one place — the turn of
// the session the work runs in — and because a row that pushed only when the
// first of them moved would go stale the moment the second did.
//
// A count rather than the questions themselves. A row prints "2 to answer" and
// nothing more, while every subscriber holds the whole project's list and is
// sent a row whenever any item changes; the questions themselves are on the
// detail, for the one item a client has open (rpc.WorkDetailSubscribeResult).
type RowState struct {
	Activity Activity
	// UnansweredQuestions is how many questions the work's session has asked
	// and nobody has answered.
	UnansweredQuestions int
}

// NeedsAttention reports whether the user is the one this work is waiting on,
// across both dimensions: a turn stuck on something only they can clear, or a
// question of the agent's that nobody has answered.
//
// This is the merge, and it happens once, here. Activity.NeedsUser is half of
// it and cannot be the whole: an agent that posted a question and carried on is
// *running*, and a predicate reading only the activity would say nobody is
// needed.
func (s RowState) NeedsAttention() bool {
	return s.Activity.NeedsUser() || s.UnansweredQuestions > 0
}

// RowStateFor derives a row's state from a work item and the turn of the
// session it runs in — the zero turn when it has no session.
func RowStateFor(w Work, turn session.TurnState) RowState {
	state := RowState{Activity: DeriveActivity(w, turn)}
	if w.Status == StatusActive || w.Status == StatusStopped {
		// A closed or never-started work draws no count. Closing withdraws the
		// questions beneath it (worktree.Manager.RetireSession), and an open one
		// has no session to have asked any — so a non-zero count on either would
		// be a leftover rather than something to act on.
		state.UnansweredQuestions = len(turn.Unanswered)
	}
	return state
}

// TurnSource resolves the turn state of every session in a worktree.
// Implemented by worktree.Manager, which answers for worktrees nobody has
// opened as well — that is the whole reason a work's activity is computed on the
// server (docs/lifecycle-ui.md §1.3).
type TurnSource interface {
	// SessionTurns returns the turn state of every session in the named
	// worktree, keyed by session id; "" names the main one. A worktree with no
	// sessions yet returns an empty map and no error.
	SessionTurns(worktree string) (map[string]session.TurnState, error)
}

// ActivityResolver derives activities for work items, reading each worktree's
// turn state at most once.
//
// It is built per batch, not kept: turn state changes with every event an agent
// produces, so a resolver that outlived the answer it was built for would hand
// out states the sessions have already left.
type ActivityResolver struct {
	source TurnSource
	turns  map[string]map[string]session.TurnState
}

// NewActivityResolver builds a resolver over one turn source. A nil source
// answers with no turns at all, which reads every active work as idle — the
// honest answer for a server that cannot see the session layer, and the shape
// narrow tests want.
func NewActivityResolver(source TurnSource) *ActivityResolver {
	return &ActivityResolver{
		source: source,
		turns:  make(map[string]map[string]session.TurnState),
	}
}

// Activity derives one work item's activity.
func (r *ActivityResolver) Activity(w Work) Activity {
	return DeriveActivity(w, r.turnFor(w))
}

// RowState derives everything one work item's row draws from the session layer.
func (r *ActivityResolver) RowState(w Work) RowState {
	return RowStateFor(w, r.turnFor(w))
}

// PendingQuestions are the questions one work item's session is waiting on
// answers to, for the detail view that lists them in full.
func (r *ActivityResolver) PendingQuestions(w Work) []session.PendingQuestion {
	return r.turnFor(w).Unanswered
}

func (r *ActivityResolver) turnFor(w Work) session.TurnState {
	if w.SessionID == "" || r.source == nil {
		return session.TurnState{}
	}
	turns, cached := r.turns[w.Worktree]
	if !cached {
		var err error
		turns, err = r.source.SessionTurns(w.Worktree)
		if err != nil {
			// A worktree that cannot be read is not a reason to blank a row: the
			// work's own status still says whether the engine is driving it, and
			// "idle" is what a work with no readable session looks like anyway.
			slog.Warn("failed to read session turns for work activity",
				"worktree", w.Worktree, "workId", w.ID, "error", err)
			turns = nil
		}
		r.turns[w.Worktree] = turns
	}
	return turns[w.SessionID]
}

File: web/src/lib/activity.ts
import {
	Circle,
	CircleCheck,
	CircleDot,
	CircleStop,
	Clock,
	Hourglass,
	Lock,
	type LucideIcon,
} from "lucide-react";
import type { SessionTurn } from "../types/message";
import type { WorkListItem, WorkStatus } from "../types/work";

/**
 * The only state any surface paints. Eight leaves, and there is never more than
 * one at a time: the layers it is derived from are each exclusive
 * (docs/lifecycle-ui.md §1.1).
 *
 * Nothing on screen may spell a state out of raw fields. A surface either
 * renders an `Activity` or it renders no state at all — which is what makes
 * this file the single place the derivation lives.
 */
export type Activity =
	| "open"
	| "running"
	| "needs_permission"
	| "background"
	| "waiting_children"
	| "idle"
	| "stopped"
	| "closed";

/** The five semantic hues an activity can wear. */
export type ActivityTone =
	| "accent"
	| "warning"
	| "secondary"
	| "muted"
	| "error";

interface ActivityView {
	Icon: LucideIcon;
	tone: ActivityTone;
	/**
	 * The state in words, for a surface with room to write it: the badge on the
	 * work detail, and the `aria-label` of a row that names its work in the same
	 * breath ("Fix the parser — Needs answer").
	 */
	label: string;
	/**
	 * What the glyph says when it stands alone, which is every surface that has
	 * no room for the label. It is a sentence rather than the label because a
	 * glyph on its own has to say who is waiting for whom — "Waiting for your
	 * answer", not "Needs answer".
	 */
	ariaLabel: string;
}

/**
 * One glyph and one tone per leaf.
 *
 * Three of these choices carry weight (docs/lifecycle-ui.md §1.1):
 * `running` and `idle` share `CircleDot` because they are the same thing — a
 * live, engine-driven work — differing only in whether a turn is open, so a
 * settling turn does not make a row appear to change identity. `background` is
 * deliberately quiet: there is nothing to do and nothing is stuck, and a louder
 * tone would re-create the two-hour spinner this redesign removes.
 * `waiting_children` keeps accent, because it is a structural state read on
 * purpose.
 *
 * `needs_permission` is the only leaf left in the warning hue, and it is the
 * whole of what an activity can say about the user being waited on. The other
 * dimension — questions waiting for an answer — is not an activity and is never
 * folded into one; see {@link needsAttention}.
 */
export const ACTIVITY_VIEW: Record<Activity, ActivityView> = {
	open: { Icon: Circle, tone: "muted", label: "Open", ariaLabel: "Open" },
	running: {
		Icon: CircleDot,
		tone: "accent",
		label: "Running",
		ariaLabel: "Agent is running",
	},
	needs_permission: {
		Icon: Lock,
		tone: "warning",
		label: "Needs permission",
		ariaLabel: "Waiting for your permission",
	},
	background: {
		Icon: Hourglass,
		tone: "secondary",
		label: "Background task",
		ariaLabel: "Waiting on a background task",
	},
	waiting_children: {
		Icon: Clock,
		tone: "accent",
		label: "Waiting on subtasks",
		ariaLabel: "Waiting on subtasks",
	},
	idle: { Icon: CircleDot, tone: "muted", label: "Idle", ariaLabel: "Idle" },
	stopped: {
		Icon: CircleStop,
		tone: "error",
		label: "Stopped",
		ariaLabel: "Stopped",
	},
	closed: {
		Icon: CircleCheck,
		tone: "muted",
		label: "Closed",
		ariaLabel: "Closed",
	},
};

/**
 * An activity as it arrived from the server, folded to one this build knows.
 *
 * Applied at the wire boundary, the way `normalizeOrigin` folds legacy message
 * origins: a client talking to a newer server must not blank a row over a leaf
 * it has never heard of, and `idle` is the leaf that claims the least.
 *
 * `Object.hasOwn` and not `in`: `in` walks the prototype chain, so `"toString"`
 * would pass for a leaf and then be looked up in the view map as a function —
 * a row rendering a glyph that does not exist. The one job of this function is
 * to not take the wire's word for it.
 */
export function normalizeActivity(raw: unknown): Activity {
	return typeof raw === "string" && Object.hasOwn(ACTIVITY_VIEW, raw)
		? (raw as Activity)
		: "idle";
}

/**
 * Whether the user is the one being waited on, across both dimensions: the one
 * activity leaf that waits on them, and the questions the session has posted and
 * nobody has answered (docs/lifecycle-ui.md §4).
 *
 * Two dimensions rather than one, because they are genuinely independent — an
 * agent that posts a question carries on running, so its activity says
 * `running` while the user still owes it an answer. Folding the question count
 * into the activity would make a running work claim to be idle.
 *
 * They are merged here and only here, for the surfaces that need a single bit:
 * the attention dot, the left edge of a row, the *Needs you* group. Anywhere a
 * user can act on one of the two, both are drawn side by side instead — a dot
 * cannot be aimed at, so it owes only the one bit.
 *
 * `background` and `waiting_children` are deliberately outside it: there is
 * nothing to do, and a dot meaning "something is happening" is a dot the user
 * learns to ignore.
 */
export function needsAttention(
	activity: Activity,
	unansweredQuestions = 0,
): boolean {
	return unansweredQuestions > 0 || activity === "needs_permission";
}

/**
 * What a client assumes before the server has said anything: no turn, and no
 * moment one started. `since` is empty rather than invented — it is only read
 * while blocked, which this never is, and a made-up timestamp would be shown.
 */
export const IDLE_TURN: SessionTurn = { phase: "idle", open: false, since: "" };

/** The part of a work this derivation reads. */
export type ActivityWork = Pick<WorkListItem, "status" | "wait">;

/**
 * Whether the engine is driving this work right now — which is the whole of
 * what `active` means. What it is waiting for while it does is `wait`, a
 * separate field, because the two are separate facts.
 */
export function isWorkActive(status: WorkStatus): boolean {
	return status === "active";
}

/**
 * The one rule (docs/lifecycle-ui.md §1.2):
 *
 * > The session says what is happening; the work's wait says what it is waiting
 * > for when nothing is happening.
 *
 * Why the phase outranks the wait rather than the other way round: a wait is a
 * standing intention, a phase is a fact about this second. An agent that calls
 * `work_wait` and then keeps writing for another ten seconds *is* running, and
 * the row should say so; the moment the turn settles the wait takes over. The
 * alternative needs a priority table between two kinds of waiting that
 * legitimately coexist, and every entry in such a table is an arbitrary choice
 * someone later "fixes".
 *
 * Permission outranks background when both are live: background is the one
 * nobody can act on, and those are the only two blockers a turn has.
 *
 * The server evaluates this same rule for work rows, which is where a row's
 * `activity` comes from (server/work/activity.go): a work list spans worktrees,
 * and a client cannot hold the turn state of a session in a worktree it has not
 * opened. The two implementations are checked against one shared table of cases,
 * `server/work/testdata/activity_cases.json`, so the rule is written down once.
 */
export function deriveActivity(
	work: ActivityWork | undefined,
	turn: SessionTurn | undefined,
): Activity {
	if (work) {
		if (work.status === "open") return "open";
		if (work.status === "closed") return "closed";
		if (work.status === "stopped") return "stopped";
	}

	if (turn?.phase === "running") return "running";
	if (turn?.phase === "blocked") {
		const kinds = turn.blockers?.map((blocker) => blocker.kind) ?? [];
		if (kinds.includes("permission")) return "needs_permission";
		return "background";
	}

	if (work?.wait === "child") return "waiting_children";
	return "idle";
}

/**
 * A session row's activity.
 *
 * The work is passed only while the engine is driving it, which is the subtle
 * call in this file: **a session row sees a work's wait, never its status.** The
 * wait is a fact about this conversation — the agent asked *here* for a message,
 * and the row is what tells the user that a session they are not looking at is
 * waiting on them. The status is not: a session outlives the work's lifecycle,
 * and a row reading `Stopped` or `Closed` would be reporting the work list's
 * business in a list that cannot act on it. Withholding the work unless it is
 * active makes those three leaves unreachable by construction.
 *
 * The caller looks the work up by the id its row carries, never by scanning for
 * a work that names the session: which sessions belong to work is the server's
 * answer now (docs/code/subscription-system.md#which-sessions-belong-to-work),
 * and the `wait` is the one thing a row still asks the work list for. So a work
 * the store has not paged in costs the row its `wait` and nothing else — it is
 * still in the right list and still says what its own turn is doing.
 */
export function sessionActivity(
	turn: SessionTurn | undefined,
	work: ActivityWork | undefined,
): Activity {
	return deriveActivity(
		work && isWorkActive(work.status) ? work : undefined,
		turn,
	);
}
